plugins/dbus_neard/dbus_neard_plugin.c

```diff
@@ -126,26 +153,145 @@ dbus_neard_plugin_name_lost(
+        /* OK to return a floating reference */
+        return g_variant_new_boolean(options->bt_static_handover);
+    } else {
+        return NULL;
+    }
+}
+
+static
+gboolean
+dbus_neard_plugin_config_set_value(
+    NfcConfigurable* config,
+    const char* key,
+    GVariant* value)
+{
+    DBusNeardPlugin* self = THIS(config);
+    DBusNeardOptions* options = &self->options;
+    gboolean ok = FALSE;
+
+    if (!g_strcmp0(key, NEARD_SETTINGS_KEY_BT_STATIC_HANDOVER)) {
+        gboolean newval = NEARD_SETTINGS_DEFAULT_BT_STATIC_HANDOVER;
+
+        if (!value) {
+            ok = TRUE;
+        } else if (g_variant_is_of_type(value, G_VARIANT_TYPE_BOOLEAN)) {
+            newval = g_variant_get_boolean(value);
+            ok = TRUE;
+        }
+
+        if (ok && options->bt_static_handover != newval) {
+            GDEBUG("%s %s", key, newval ? "on" : "off");
+            options->bt_static_handover = newval;
+            g_signal_emit(self, dbus_neard_plugin_signals
+                [SIGNAL_CONFIG_VALUE_CHANGED], g_quark_from_string(key),
+                key, value);
+        }
+    }
+    return ok;
+}
+
+static
+gulong
+dbus_neard_plugin_config_add_change_handler(
+    NfcConfigurable* config,
+    const char* key,
+    NfcConfigChangeFunc func,
+    void* user_data)
+{
+    return g_signal_connect_closure_by_id(THIS(config),
+        dbus_neard_plugin_signals[SIGNAL_CONFIG_VALUE_CHANGED],
+        key ? g_quark_from_string(key) : 0,
+        g_cclosure_new(G_CALLBACK(func), user_data, NULL), FALSE);
+}
+
+static
+void
+dbus_neard_plugin_config_init(
+    NfcConfigurableInterface* iface)
+{
+    iface->get_keys = dbus_neard_plugin_config_get_keys;
+    iface->get_value = dbus_neard_plugin_config_get_value;
+    iface->set_value = dbus_neard_plugin_config_set_value;
+    iface->add_change_handler = dbus_neard_plugin_config_add_change_handler;
+}
+
+/*==========================================================================*
+ * NfcPlugin
+ *==========================================================================*/
+
 static
 gboolean
 dbus_neard_plugin_start(
     NfcPlugin* plugin,
     NfcManager* manager)
 {
-    DBusNeardPlugin* self = DBUS_NEARD_PLUGIN(plugin);
+    const char* legacy_config_file = NEARD_SETTINGS_FILE;
+    DBusNeardPlugin* self = THIS(plugin);
+    DBusNeardOptions* options = &self->options;
     NfcAdapter** adapters;
 
     GVERBOSE("Starting");
+
+    /* Migrate the config */
+    if (g_file_test(legacy_config_file, G_FILE_TEST_EXISTS)) {
+        GKeyFile* keyfile = g_key_file_new();
+
+        GINFO("Migrating %s", legacy_config_file);
+        if (g_key_file_load_from_file(keyfile, legacy_config_file, 0, NULL)) {
+            GError* error = NULL;
+            const gboolean value = g_key_file_get_boolean(keyfile,
+                NEARD_SETTINGS_GROUP, NEARD_SETTINGS_KEY_BT_STATIC_HANDOVER,
+                &error);
+
+            if (!error) {
+                options->bt_static_handover = value;
```

Comment:
Presumably this gets saved in non-legacy place, or?
When does that happen? Should one worry about losing settings e.g. due to crash between legacy file removal below and reaching such save data position?

Reply:
Yes, it's saved in a new place (note to myself - double check that) and yes, crashing nfcd between this place where the old file is removed and the place where the new storage gets updated (which happens pretty soon - after all plugins have started) would lose the last configured value and reset it back to default (if it was different from the default). The probability of that to happens is really really low, almost zero.

I decided that making it super-duper reliable wasn't worth the trouble.

Reply:
I actually got the legacy file name wrong in the original version of the pull request 😮. Fixed the file name, tested the migration, it worked.